docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,546 @@
+        if (!compressionDisabled && (streamingPayload || payloadSize!! >= compressionThreshold)) {
+
+            // Check if requested algorithm(s) is supported
+            findAlgorithm()?.let { algorithm ->
+
+                // Attempt compression
+                val request = context.protocolRequest.toBuilder()
+                request.body = request.body
+                // TODO: Add compression for streams
+                addHeader(request, algorithm.algorithmId())
+                return request.build()
+            }
+        }
+        return context.protocolRequest
+    }
+}
+```
+
+Supporting function to help find if requested algorithm is supported:
+```kotlin
+/**
+ * Finds first match in supported and request
+ */
+private fun findAlgorithm() : CompressionAlgorithm? {
+    requestedCompressionAlgorithms.forEach { requestedAlgorithmAlgorithmId ->
+        supportedCompressionAlgorithms.forEach { supportedAlgorithm ->
+            if (supportedAlgorithm.algorithmId() == requestedAlgorithmAlgorithmId) return supportedAlgorithm
+        }
+    }
+    return null
+}
+
+// TODO: Return list instead of single algorithm
+
+```
+
+Supporting function to add header or modify header once compression is successful:
+```kotlin
+/**
+ * Appends the algorithm id to the content encoding header. Doesn't remove old content encodings if already present
+ * in header
+ */
+private fun addHeader(request: HttpRequestBuilder, algorithmId: String) {
+    val previousEncodings = request.headers["Content-Encoding"]
+    val contentEncodingHeaderPrefix = previousEncodings?.let { "$previousEncodings, " } ?: ""
+
+    request.headers.remove("Content-Encoding")
+    request.header(
+        "Content-Encoding",
+        "$contentEncodingHeaderPrefix${algorithmId}"
+    )
+}
+```
+
+## Client Experience
+The client experience should be seamless unless they are doing something wrong.
+
+Here is how a client might use/configure this:
+```kotlin
+val client = S3Client {
+    region = "us-west-2"
+    disableRequestCompression = true
+    requestMinCompressionSizeBytes = 200
+}
+```
+or
+```kotlin
+val client = S3Client.fromEnvironment {
+    region = "us-west-2"
+}
+```
+```kotlin
+// In config file 
+[default]
+disable_request_compression = true
+request_min_compression_size_bytes = 200
+
+```
+Both of these examples would achieve the same result.
+
+## Testing
+Testing should be done using both unit tests and Smithy HTTP protocol compliance tests. The  SDK unit tests will be used to verify behaviour and ensure it doesn't drift over time (unless intended to). And each AWS HTTP protocol test will be used to test the behavior in a more integrated way in comparison to the unit tests.
+
+AWS JSON 1.0 smithy protocol test for `requestCompression` trait:
+```json
+$version: "2.0"
+
+namespace aws.protocoltests.json10
+
+        use aws.protocols#awsJson1_0
+        use smithy.test#httpRequestTests
+        use smithy.test#httpResponseTests
```

Comment:
**Correctness**: This appears to be incompletely copied/pasted from the specification. Is there any difference in what you want to show here? If not then there's little reason to duplicate it from the spec.